crates/autopilot/src/domain/settlement/observer.rs

```diff
@@ -29,80 +30,61 @@ pub struct Observer {
+    pub async fn post_process_outstanding_settlement_transactions(&self) {
+        let settlements = Self::retry_with_sleep(|| async move {
+            self.persistence
+                .get_settlements_without_auction()
+                .await
+                .inspect_err(|err| {
+                    tracing::warn!(?err, "failed to fetch unprocessed settlements from DB")
+                })
+        })
+        .await
+        .unwrap_or_default();
 
-            // everything worked fine -> reset our attempts for the next settlement
-            attempts = 0;
+        if settlements.is_empty() {
+            tracing::debug!("no unprocessed settlements found");
+            return;
         }
-    }
-
-    /// Update database for settlement events that have not been processed yet.
-    ///
-    /// Returns whether an update was performed.
-    async fn single_update(&self) -> Result<IndexSuccess> {
-        // Find a settlement event that has not been processed yet.
-        let Some(event) = self
-            .persistence
-            .get_settlement_without_auction()
-            .await
-            .context("failed to fetch unprocessed tx from DB")?
-        else {
-            return Ok(IndexSuccess::NothingToDo);
-        };
 
-        tracing::debug!(tx = ?event.transaction, "found unprocessed settlement");
+        // On mainnet it's common to have multiple settlements in the
+        // same block. So even if we process every block immediately,
+        // we should still post-process multiple settlements concurrently.
+        const MAX_CONCURRENCY: usize = 10;
+        futures::stream::iter(settlements)
+            .for_each_concurrent(MAX_CONCURRENCY, |settlement| async move {
+                tracing::debug!(tx = ?settlement.transaction, "start post processing of settlement");
+                match Self::retry_with_sleep(|| self.post_process_settlement(settlement)).await {
+                    Some(_) =>  tracing::debug!(tx = ?settlement.transaction, "successfully post-processed settlement"),
+                    None => tracing::warn!(tx = ?settlement.transaction, "gave up on post-processing settlement"),
+                }
+            })
+            .await;
+    }
 
+    async fn post_process_settlement(&self, settlement: eth::SettlementEvent) -> Result<()> {
         let settlement_data = self
-            .fetch_auction_data_for_transaction(event.transaction)
+            .fetch_auction_data_for_transaction(settlement.transaction)
             .await?;
         self.persistence
-            .save_settlement(event, settlement_data.as_ref())
+            .save_settlement(settlement, settlement_data.as_ref())
             .await
-            .context("failed to update settlement")?;
-
-        match settlement_data {
-            None => Ok(IndexSuccess::SkippedInvalidTransaction),
-            Some(_) => Ok(IndexSuccess::IndexedSettlement),
-        }
+            .context("failed to update settlement")
```

Comment:
Given that we're running multiple ones in parallel, won't this apply extra write pressure on the DB?

Should we rather aggregate concurrent reads and write a single batch?

Reply:
There is very little data we write to the DB here. Before making the code more complicated here I'd actually want to see that unbatched writes are causing issues.
Also before this PR the writes have also been unprocessed. The difference is with the parallel execution there is at least a chance that they hit quickly after another (instead of being spread out due to the serial processing).